pkg/loop/internal/pb/chain_reader.proto

```diff
@@ -9,9 +9,112 @@ import "google/protobuf/empty.proto";
 
 service ChainReader {
   rpc GetLatestValue (GetLatestValueRequest) returns (GetLatestValueReply) {}
+  rpc QueryKeys (QueryKeysRequest) returns (QueryKeysReply) {}
   rpc Bind(BindRequest) returns (google.protobuf.Empty) {}
 }
 
+message AndFilter {
+  repeated QueryKeysFilter filters = 1;
+}
+
+message AddressFilter {
+  repeated string addresses = 1;
+}
+
+message KeysFilter {
+  repeated string keys = 1;
+}
+
+message KeysByValueFilter {
+  repeated Keys keys = 1;
+  repeated Values values = 2;
+
+}
+
+message Keys {
+  repeated string keys = 1;
+}
+
+message Values {
+  repeated string values = 1;
+}
+
+message Confirmations {
+  int32 confirmation = 1;
+}
+
+message ConfirmationsFilter{
+  Confirmations confirmations = 1;
+}
+
+message ComparisonOperator {
+  int32 comparisonOperator = 1;
+}
+
+message BlockFilter{
+  uint64 block_number = 1;
+  ComparisonOperator operator = 2;
+}
+
+message TimestampFilter{
+  uint64 timestamp = 1;
+  ComparisonOperator operator = 2;
+}
+
+message TxHashFilter{
+  string tx_hash = 1;
+}
+
+enum SortDirection {
+  Asc = 0;
+  Desc = 1;
+}
+
+message SortByTimestamp {
+  SortDirection sort_direction = 1;
+}
+
+message SortByBlock {
+  SortDirection sort_direction = 1;
+}
+
+message SortBySequence {
+  SortDirection sort_direction = 1;
+}
+
+message SortBy {
+  oneof sort_by {
+    SortByTimestamp sort_by_timestamp = 1;
+    SortByBlock sort_by_block = 2;
+    SortBySequence sort_by_sequence = 3;
+  }
+}
+
+message LimitAndSort{
+  uint64 limit = 1;
+  repeated SortBy sort_by = 2;
+}
+
+message QueryKeysFilter {
+  oneof filter {
+    AndFilter and_filter = 1;
```

Comment:
i'm i reading this correctly: it's an infinite recursion?

QueryKeysFilter -> AndFilter -> QueryKeysFilter -> ...

can you remove the AndFilter and have something like ~QueryKeysFilter {
// repeated filters are AND'd
repeated filters {
	oneof  {
	AddressFilter
	...
	}
	
	not sure what the syntax is supposed to be

Reply:
It just means that we can have nested ands. This doesn't enter recursion anywhere or break anything



`

Reply:
it's going to be hard to use and understand because it's mixing primitive operators and derived concepts into the same thing.

Reply:
I've looked at some orms, and this is similar to how it usually works, except that I'm not allowing arbitrary data to be used with conditions(=,>=,!=...) but predefined conditions(block,timestamp ...)

Reply:
ok, i think i get the idea. i would have been tempted to do ~

```
message QueryFilter {
repeated Expr expr
}


message Expr {
oneof evualator {
Primitive primitive
Composite composite}
}

// first class primitives for comparing chain data
message Primitive {
	oneof compatior {
	AddressFilter address_filter = 3; // hard coded to check for equality
    ConfirmationsFilter confirmations_filter = 4;
    BlockFilter block_filter = 5; // allows comparision with eq, neq,...
    TimestampFilter timestamp_filter = 6;
    TxHashFilter tx_hash_filter = 7;
}
}

//composite enables us to chain  exprs
message Composite {
 oneof comparitor {
 And and
 Or or
}
}

// and checks the both exprs are true
message And {
  Expr arg1
  Expr arg2
}

// or checks at either expr is true
message Or {
 Expr arg1
 Expr arg2
}
```

i believe this expresses the same thing as what you have. it's easier for me to think it terms of these seperated building blocks. i think it would read in more or less the same way you've written the example go code

Reply:
Its really hard to make this intuitive and also parse-able at the end destination.

I provided an alternative here https://github.com/smartcontractkit/chainlink-common/commit/73ef07d6a97e0f91a73109f12afb1ffcf85d6dd5

Reply:
It all ends up looking very similar from the perspective of the end user using functions to build out the query.